File: src/wrappers/erlang/erlpiphany/examples/hello-world/src/hello_world.c
/*
  hello_world.c

  Copyright (C) 2012 Adapteva, Inc.
  Contributed by Yaniv Sapir <yaniv@adapteva.com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program, see the file COPYING.  If not, see
  <http://www.gnu.org/licenses/>.
*/

// This is the HOST side of the Hello World example.
// The program initializes the Epiphany system,
// randomly draws an eCore and then loads and launches
// the device program on that eCore. It then reads the
// shared external memory buffer for the core's output
// message.

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <e-hal.h>

#define _BufSize   (128)
#define _BufOffset (0x01000000)
#define _SeqLen    (20)

int main(int argc, char *argv[])
{
	unsigned row, col, coreid, i;
	e_platform_t platform;
	e_epiphany_t dev;
	e_mem_t emem;
	char emsg[_BufSize];

	srand(1);

	// initialize system, read platform params from
	// default HDF. Then, reset the platform and
	// get the actual system parameters.
	e_init(NULL);
	e_reset_system();
	e_get_platform_info(&platform);

	// Allocate a buffer in shared external memory
	// for message passing from eCore to host.
	e_alloc(&emem, _BufOffset, _BufSize);

	for (i=0; i<_SeqLen; i++)
	{
		// Draw a random core
		row = rand() % platform.rows;
		col = rand() % platform.cols;
		coreid = (row + platform.row) * 64 + col + platform.col;
		fprintf(stderr, "%3d: Message from eCore 0x%03x (%2d,%2d): ", i, coreid, row, col);

		// Open the single-core workgroup and reset the core, in
		// case a previous process is running. Note that we used
		// core coordinates relative to the workgroup.
		e_open(&dev, row, col, 1, 1);
		e_reset_group(&dev);

		// Load the device program onto the selected eCore
		// and launch after loading.
		e_load("e_hello_world.srec", &dev, 0, 0, E_TRUE);

		// Wait for core program execution to finish, then
		// read message from shared buffer.
		usleep(10000);
		e_read(&emem, 0, 0, 0x0, emsg, _BufSize);

		// Print the message and close the workgroup.
		fprintf(stderr, "\"%s\"\n", emsg);
		e_close(&dev);
	}

	// Release the allocated buffer and finalize the
	// e-platform connection.
	e_free(&emem);
	e_finalize();

	return 0;
}

File: src/wrappers/erlang/erlpiphany/examples/hello-world/src/hello_world.erl
%%%-------------------------------------------------------------------
%%% @author Mark A Fleming <mark_fleming@ieee.org>
%%% @copyright (C) 2015, Erlang Public License
%%% @doc
%%% This module duplicates the functionality of the Adapteva, Inc.
%%% C program of the same name, by Yaniv Sapir.
%%% 
%%% To quote the original program documentation
%%% 
%%% ```// This is the HOST side of the Hello World example.
%%% // The program initializes the Epiphany system,
%%% // randomly draws an eCore and then loads and launches
%%% // the device program on that eCore. It then reads the
%%% // shared external memory buffer for the core's output
%%% // message.'''
%%% 
%%% Make sure the erlpiphany include, ebin and priv directories are in
%%% your path or use/modify the init function below.
%%% @end
%%% Created : 21 Jan 2015 by Mark A Fleming <mark_fleming@ieee.org>
%%%-------------------------------------------------------------------
-module(hello_world).

%%-on_load(init/0).

-export([main/0, init/0]).

-define(BufSize, (128)).
-define(BufOffset, (16#01000000)).
-define(SeqLen, 20).


init() ->
    %% Add (relative) path to appropriate directories
    code:add_pathz("../../../include"),
    code:add_pathz("../../../priv"),
    code:add_pathz("../../../ebin").

main() ->
    {A1, A2, A3} = now(),
    random:seed(A1, A2, A3),
    
    %% initialize system
    ok = ehal:e_init(0),
    ok = ehal:e_reset_system(),
    {ok, Platform} = ehal:e_get_platform_info(),

    %% Allocate a buffer in shared memory
    {ok, Emem} = ehal:e_alloc(?BufOffset, ?BufSize),

    %% Process, recurse not loop
    loop(Platform, Emem, 0, ?SeqLen-1),

    %% Release buffer and finalize
    ok = ehal:e_free(Emem),
    ok = ehal:e_finalize().



loop(Plat, Emem, N, N) ->
    process(Plat, Emem, N);
loop(Plat, Emem, I, N) ->
    process(Plat, Emem, I),
    loop(Plat, Emem, I+1, N).



process(Plat, Emem, I) ->
    %% Platform info is a list of key-value tuples where key is the
    %% structure member name
    {row, Row} = lists:keyfind(row, 1, Plat),
    {col, Col} = lists:keyfind(col, 1, Plat),
    {rows, Rows} = lists:keyfind(rows, 1, Plat),
    {cols, Cols} = lists:keyfind(cols, 1, Plat),

    %% Draw a random core
    R = random:uniform(Rows)-1,
    C = random:uniform(Cols)-1,
    Coreid = (R+Row)*64 + C + Col,
    io:format("~3w: Message from eCore ~.16x (~2w,~2w): ", [I, Coreid, "0x", R, C]),

    %% Open single-core workgroup
    {ok, Dev} = ehal:e_open(R, C, 1, 1),
    ok = ehal:e_reset_group(Dev),

    %% Load program to selected core
    ok = ehal:e_load("e_hello_world.srec", Dev, 0, 0, true),

    %% Wait for program to finish
    timer:sleep(10),
    {ok, {?BufSize, Data}} = ehal:e_read(Emem, 0, 0, 0, ?BufSize),

    %% Print message, close workgroup
    Msg = binary_to_list(Data),
    String = lists:filter(fun(E) ->
				     E>0
			     end, Msg),
    io:format("~p~n", [String]),
    ok = ehal:e_close(Dev).
